spark/src/test/scala/org/apache/spark/sql/delta/rowid/RowTrackingCompactionSuite.scala

```diff
@@ -0,0 +1,305 @@
+          .select(col("id") % 10 as "key", col("id") as "value")
+        writeDf(dir, partitioned, withMaterializedRowTrackingColumns, df)
+      }
+      // Write 3 times to create 3 commits with different versions
+      writeValues(start = 0, end = 100)
+      writeValues(start = 100, end = 200)
+      writeValues(start = 200, end = 300)
+    }
+  }
+
+  protected def writeDf(
+      dir: File,
+      partitioned: Boolean,
+      withMaterializedRowTrackingColumns: Boolean,
+      _df: DataFrame): Unit = {
+    var df = _df
+    if (withMaterializedRowTrackingColumns) {
+      val deltaLog = DeltaLog.forTable(spark, dir)
+      val snapshot = deltaLog.update()
+      val materializedRowIdColName = MaterializedRowId.getMaterializedColumnNameOrThrow(
+        snapshot.protocol, snapshot.metadata, deltaLog.tableId)
+      df = df.withMaterializedRowIdColumn(materializedRowIdColName, col("value"))
+      val materializedRowCommitVersionColName =
+        MaterializedRowCommitVersion.getMaterializedColumnNameOrThrow(
+          snapshot.protocol, snapshot.metadata, deltaLog.tableId)
+      df = df.withMaterializedRowCommitVersionColumn(
+        materializedRowCommitVersionColName, col("value"))
+    }
+
+    var writer = df.write.format("delta").mode("append")
+    if (partitioned) {
+      writer = writer.partitionBy("key")
+    }
+    writer.save(dir.getAbsolutePath)
+  }
+
+  protected def runStatement(statement: String): OptimizeMetrics = {
+    import testImplicits._
+    spark.sql(statement)
+      .select(col("metrics.*"))
+      .as[OptimizeMetrics]
+      .head()
+  }
+
+  protected def runCompaction(dir: File, applyFilter: Boolean): OptimizeMetrics
+
+  protected def checkCompactionRowTrackingPreservation(dir: File, applyFilter: Boolean): Unit = {
+    val deltaLog = DeltaLog.forTable(spark, dir.getAbsolutePath)
+
+    def readWithRowTrackingColumns(): DataFrame = {
+      spark.read.format("delta").load(dir.getAbsolutePath)
+        .select("value", RowId.QUALIFIED_COLUMN_NAME, RowCommitVersion.QUALIFIED_COLUMN_NAME)
+    }
+
+    val rowsBeforeFirstCompaction = readWithRowTrackingColumns().collect()
+
+    // Run compaction, check that at least one file in the table was rewritten, and check that
+    // the fresh row IDs have been preserved.
+    checkRowTrackingMarkedAsPreservedForCommit(deltaLog) {
+      val metrics = runCompaction(dir, applyFilter)
+      assert(metrics.numFilesRemoved > 0)
+    }
+    checkAnswer(readWithRowTrackingColumns(), rowsBeforeFirstCompaction)
+
+    assertRowIdsAreValid(deltaLog)
+  }
+
+  protected def runTest(
+      partitioned: Boolean,
+      withMaterializedRowTrackingColumns: Boolean,
+      applyFilter: Boolean): Unit = {
+    withTempDir { dir =>
+      createTable(dir, rowTrackingEnabled = true, partitioned, withMaterializedRowTrackingColumns)
+      checkCompactionRowTrackingPreservation(dir, applyFilter)
+    }
+  }
+
+  protected def commandName: String
```

Comment:
Pull this at the top of this trait, it's hard to see that this should be overriden by derived classes otherwise